### ISSUE TITLE
Baker Analytics BakerIssueClose called upon open (and again upon close)

### ISSUE THREAD

**Issue by bdgeyter** [Simbul/baker#1144]
When you open an issue, the BakerViewController `viewWillDisappear:` method is called and the `BakerIssueClose` notification is posted right after opening the issue. I'm not entirely sure why it's called there, but a fix would be to move the `postNotificationName:` method to

```
- (void)willMoveToParentViewController:(UIViewController *)parent{
     [[NSNotificationCenter defaultCenter] postNotificationName:@"BakerIssueClose" object:self]; // -> Baker Analytics Event 
}
```

**Comment by folletto** [comment]
Seems like a good suggestion, we'll look into it. :)

**Comment by bdgeyter** [comment]
Additionally, `BakerViewPage` is called in BakerViewController `loadSlot:WithPage:`. This works fine for the very first page, but every page after that (swipe or tap) is not logged.

To track additional pages, the notification should also be posted in the `goToPage` method.

**Comment by folletto** [comment]
Uh that's weird, in the tests it worked. We'll review that as well, thanks. :)

**Comment by bdgeyter** [comment]
Maybe it depends on the difference between baker-rendering modes? 

My guess is (without test) if you use screenshot mode, it will always call `loadSlot:WithPage` with the `currentPageNumber`, and post the notification, but in three-cards mode (which I use) it prerenders pages, and no notification is posted.

**Comment by fturner19** [comment]
@folletto - I'm experiencing the same issue upon further testing with page counts! @bdgeyter i'll try your edit and see if i get the results i'm looking for!

**Comment by folletto** [comment]
Ah yes @bdgeyter, good point. :) Well spotted! Thanks. :)

**Comment by MrKistic** [comment]
Good work @bdgeyter, I ran into exactly the same issues and the `willMoveToParentViewController` solution works well for me. @folletto I'd like to point out that `viewWillDisappear:` also gets called when `loadModalWebView:` is called, meaning that you get `BakerIssueClose` tracked whenever you click an internal link!

I also discovered the issue with `loadSlot` and stuck a new Analytics event into `goToPage:`. I'm just testing that at the moment but it seems to work pretty well for me.

I wish I'd found this issue first though, that would have saved a bit of time today...

**Comment by MrKistic** [comment]
Some further notes here, currently both the `BakerIssuePurchase` and `BakerSubscriptionPurchase` analytics events are called as soon as the user pushes either the Subscribe or Buy button. They may immediately press Cancel on the iTunes dialogue and do no such thing but we're still recording the event in GA.

Better would be to make the calls in the `PurchasesManager` `-completeTransaction:` method, I think. At least this way they're only called when the process has actually been completed. That seems to be working for me so far anyway.

Unfortunately it will also be called still if somebody has already purchased it and restores it... but I'm not sure that I can do much about that without diving a lot further into that process.

**Comment by folletto** [comment]
This last one is actually done on purpose, to track something that's not tracked. You see, both actions are already tracked elsewhere: either in iTunes Connect or your backend server (or both). Given the nature of analytics softwares, including GA, it's likely there will be a mismatch between the two values, but just iTunes/server will be accurate.

Instead, what's not tracked and is very valuable, is exactly the press of the button. This allows you to discover things by matching the two metrics. Things like: "Why I have 100 presses of the download button but just 10 sales on the backend?".

In the best case scenario, the two values should match. But if they don't, then you have something to think about and work on. ;)

**Comment by MrKistic** [comment]
Yes of course, that makes a lot of sense. In fact if this was a web app I was building, I would definitely be tracking both events as well as maintaining my purchase records (i.e. iTunes/server) and comparing the lot of them.

Which is exactly what I'm going to do :)

Thanks again.

**Comment by Xm4s** [comment]
The wrong dispatch of the events `BakerIssueClose` and `BakerViewPage` has been addressed in the commits ef444b762c441500762d80906d644097f1690b4e and c427d4a3c737852810dd953478c9e302acdfad0b, now everything should be working properly.

Thanks to @bdgeyter and @MrKistic for the help

**Comment by MrKistic** [comment]
Ah yes, just after I upload an update to the app store!

I did go with @bdgeyter's earlier suggestion for `BakerIssueClose` which seems a bit cleaner:

```
- (void)willMoveToParentViewController:(UIViewController *)parent{
  if (self.isViewLoaded) { // this is also called when loading - only trigger the event on unloading
    [[NSNotificationCenter defaultCenter] postNotificationName:@"BakerIssueClose" object:self]; // -> Baker Analytics Event
    }
}
```

I didn't think to move the `saveBookStatusWithScrollIndex` call but that makes sense. Anyway thanks for the update, happy to have helped.
